Add function process_fci() for rutokenS

Mardalemer opened the pull request.

frankmorgner committed:
pkcs15-tool: added test for session PIN

frankmorgner committed:
reader-pcsc: removed cardmod driver

- pcsc driver takes over all the functionality
- no dedicated reader driver config values for cardmod, use application
  specific blocks to define a different behavior for the pcsc reader if
  needed
- removes legacy code; requiring at least libpcsclite 1.6.5

Fixes https://github.com/OpenSC/OpenSC/issues/892

fabled committed:
pkcs15init: use flag symbolic names

fabled committed:
pkcs15: fix and simplify object type class comparisons

frankmorgner committed:
md: added cancel button to PINPAD GUI

... as required by minidriver spec

- Shows PIN purpose depending on the PIN type similar to Base CSP

frankmorgner committed:
md: Implemented German locale for PINPAD-dialog

frankmorgner committed:
macOS: added support for access via CryptoTokenKit

Binaries still need to be signed with the com.apple.security.smartcard
entitlement. The command should look something like this:

codesign --force --entitlements MacOSX/pcsc.entitlements --sign "Mac Developer" target/Library/OpenSC/bin/*
codesign --force --entitlements MacOSX/pcsc.entitlements --sign "Mac Developer" target/Library/OpenSC/lib/*.dylib
codesign --force --entitlements MacOSX/pcsc.entitlements --sign "Mac Developer" --deep target/Library/OpenSC/lib/opensc-pkcs11.bundle
codesign --force --entitlements MacOSX/pcsc.entitlements --sign "Mac Developer" --deep target/Library/Security/tokend/OpenSC.tokend

frankmorgner committed:
md: added themable PIN-pad dialog

- themable in the sense of using OS native design
- user messages on PIN pad dialog are identical to Base CSP, which still displays the dialog for PIN entry if no PIN pad is available
- adds progress bar to dialog
- Uses Smartcard icon extracted from DDORes.dll
- requires windows vista/windows server 2008 or above

frankmorgner committed:
md: fixed race condition when showin pin pad dialog

When the dialog is shown in a separate thread and the user removes the
card, both, the thread for the pin pad operation and the main thread
are trying to access the card and context handles. Even worse, the main
thread deletes the context handle, which may result in a segmentation
fault for the thread with the pin pad operation.

frankmorgner committed:
md: Use card specifiic pin pad strings

implemented for sc-hsm SoC

frankmorgner committed:
md: configure certificate propagation service

frankmorgner committed:
replace assert with error handling

frankmorgner committed:
md: Delegate user to the PIN pad's "Cancel" button

Pressing the cancel button in the PIN pad dialog should not close the
dialog. The application will still wait for the request to complete
even if the dialog is gone. Instead, we tell the user to press the
cancel butten on the PIN pad if the reader does not support SCardCancel.

frankmorgner committed:
fixed CertPropSvc configuration on uninstallation

frankmorgner committed:
minidriver: added support for session pin

frankmorgner committed:
md: use c89 style to silence VC2010

frankmorgner committed:
win32 installer: skip license agreement

in accordance to
https://www.gnu.org/licenses/old-licenses/gpl-2.0-faq.html#ClickThrough

dengert committed:
PKCS#11 does not define a CKA_VALUE for public keys and is missused

OpenSC opennssl.c in sc_pkcs11_verify_data assumes that it can
retieve the CKA_VALUE for a public key object, and expect it to
be usable as RSA.

But internally sc_pkcs15_pubkey can have a "raw" or "spki"
version of the public key as defined by PKCS#15.  Card drivers
or pkcs15-<card> routines may store either the "raw" or "spki"
versions. A get attribute request for CKA_VALUE for a public key
will return either the raw, spki or will derived rsa verison of the
pubkey.

This commit will test if the CKA_VALUE is a spki and use d2i_PUBKEY
which takes a spki version and returns an EVP_KEY. If it not an spki
the current method, d21_PublicKey(EVP_PKEY_RSA,...) is used which
only works for RSA.

The problem was found while testing pkcs11-tool -t -l  where
the  verify tests would fail with a CKR_GENERAL_ERROR because
the card driver stored the public key as a spki.

On branch verify-pubkey-as-spki-2
 Changes to be committed:
	modified:   src/pkcs11/openssl.c

Date:      Fri Apr 07 07:50:00 2017 -0600

frankmorgner committed:
sc-hsm: Add support for SoC

- eac: allow CA without EF.CardSecurity
- sc-hsm: implemented CA based on document PKI
- sc-hsm: adds receive limit for SoC card
- introduces dedicated card type for SoC card
- md: integrate card's PIN pad capabilities
- installer: added SC-HSM SoC card to registry
- pkcs15-tool: Added support for PIN entry on card
- change/unblock PIN: add support for PIN entry on card
- added OpenPACE to macOS build
- travis-ci: install gengetopt/help2man via brew
- sc-hsm: Cache EF.C_DevAut
- sc-hsm: Prevent unnecessary applet selection and state resets
- sc-hsm: added support for session pin
- sc-hsm: avoid multiple AID selection
- sc-hsm: Use the information from match_card for all subsequent selections of the applet
- sc-hsm: cache optional files as empty files (Decoding the files will reveal that they were not existing prior caching. This avoids selecting the file though we have already tried to cache the file before.)
- use dedicated directory for CVC trust anchors
- appveyor: added OpenPACE to windows build

frankmorgner committed:
fixed import of multiple symbols

regression of 45a7ea9737075b5901fe7a5d65ed898733140315:
due to the change in the linkage, the symbols should be found in
opensc.dll instead of the static support libraries.

frankmorgner committed:
recognize T=0 limitation of sending 255 bytes

fixes https://github.com/OpenSC/OpenSC/issues/1009

frankmorgner committed:
pkcs11-tool: fixed RSA-X-509 corner case test

For testing RSA-X-509, we are generating random bytes for signing. It
may happen that the modulus is smaller than the random number
generated, which triggers an error in the card. With this change, we
are setting the most significant byte to 0x00 to assure the random
number is smaller than the modulus.

frankmorgner committed:
Coverity fixes (#1012)

card-cac.c
 * CLANG_WARNING: The left operand of '<' is a garbage value
card-coolkey.c
 * CLANG_WARNING: overwriting variable
 * CPPCHECK_WARNING: memory leak / overwrite variable
 * CLANG_WARNING: null pointer dereference
 * UNUSED_VALUE: unused return value
card-gids.c
 * CLANG_WARNING: Branch condition evaluates to a garbage value
 * SIZEOF_MISMATCH: suspicious_sizeof
card-myeid.c
 * RESOURCE_LEAK: Variable "buf" going out of scope leaks the storage it points to.
 * CLANG_WARNING: overwriting variable
 * (rewrite not to confuse coverity)
pkcs15-cac.c
 * RESOURCE_LEAK: Variable "cert_out" going out of scope leaks the storage it points to.
pkcs15-coolkey.c
 * UNUSED_VALUE: unused return value
pkcs15-piv.c
 * RESOURCE_LEAK: Variable "cert_out" going out of scope leaks the storage it points to.
pkcs15-sc-hsm.c
 * DEADCODE
pkcs11/framework-pkcs15.c
 * RESOURCE_LEAK: Variable "p15_cert" going out of scope leaks the storage it points to.
pkcs15init/pkcs15-lib.c
 * CLANG_WARNING: Assigned value is garbage or undefined
pkcs15init/pkcs15-myeid.c
 * UNREACHABLE: Probably wrong placement of code block
tests/p15dump.c
 * IDENTICAL_BRANCHES
pkcs15-init.c
 * CLANG_WARNING: Potential leak of memory pointed to by 'args.der_encoded.value'
pkcs15-tool.c
 * RESOURCE_LEAK: Variable "cert" going out of scope leaks the storage it points to.
 * MISSING_BREAK: The above case falls through to this one.
sc-hsm-tool.c
 * CLANG_WARNING: Potential leak of memory pointed to by 'sp'
westcos-tool.c
 * FORWARD_NULL: Passing null pointer "pin" to "unlock_pin", which dereferences it.
 * (rewrite not to confuse coverity)
card-cac.c
* Avoid malloc with 0 argument
gids-tool.c
* FORWARD_NULL -- copy&paste error
scconf.c
* CLANG_WARNING: Call to 'malloc' has an allocation size of 0 bytes

closes #982

frankmorgner committed:
Avoid dnie_transmit_apdu in the dnie driver (#970) (#1013)

closes #975 
closes #972

Jakuje committed:
Make CardOS 5.3 working with OpenSC (#1003)

* Set security context for CardOS 5.3 with p1=0x41 (as Coolkey does)

* Do not emulate signatures in CardOS 5.3

Remove the bogus SC_ALGORITHM_NEED_USAGE which prevents using the
actual implementation in cardos_compute_signature().

It might be bogus also in previous version, but I don't have a way
to verify against these cards.

* Do not advertise RSA-X-509 mechanism for CardOS 5.3 (card strips padding)

hhonkanen committed:
Added a check to sc_pkcs15_verify_pin to find out if the access condition is already open on card. This check is performed only if this function is called with empty data. This change fixes a problem with pinpad readers, when PIN cache is disabled and prevents unnecessary PIN queries.

hhonkanen committed:
Improved creation of key files so that the correct security attributes are set and keys can be created under specific PINs. Previously keys were always created under PIN 1. Changed description of myeid_create_key function.

dengert committed:
Add new attribute CKA_SPKI for CKO_PUBLIC_KEY

CKA_SPKI is a vendor defined attribute to be used internally
as input to to OpenSSL d2i_PUBKEY

 On branch verify-pubkey-as-spki-2
 Changes to be committed:
	modified:   framework-pkcs15.c
	modified:   mechanism.c
	modified:   openssl.c
	modified:   pkcs11-opensc.h

dengert committed:
Make sure it is an SPKI

 On branch verify-pubkey-as-spki-2
 Changes to be committed:
	modified:   framework-pkcs15.c

hamano committed:
fix trivial leak

fabled committed:
Add _sc_card_add_generic for registering secret key algorithms

fabled committed:
pkcs15init: generalize key compatibility checking helpers

fabled committed:
pkcs15init: handle secret key in generic helper functions

Jakuje committed:
Make sure the lock is released when returning

Jakuje committed:
pkcs11-tool: Do not use unitialized data when C_GetTokenInfo() failed

Jakuje committed:
pkcs11-tool: Set CKA_PRIVATE=false for new public keys

Keygen should write public keys with explicit CKA_PRIVATE=false by default (possibility to modify by --private switch)

Related to 4df35b92 discussing writing separate objects years ago.

frankmorgner committed:
fixed more coverity issues

frankmorgner committed:
removed obsolete comment

fabled committed:
add AES algorithm ID

fabled committed:
Imporve SKDF decoding and implement encoding

- fixes decoding of SecretKeyAttributes
- adds support for algorithmReferences
- adds support for algIndependentKeys (PKCS#15 Generic keys)
- implements encoding of SKDF

fabled committed:
pkcs15-tool: fix secret key dumping

fabled committed:
pkcs15init: prepare sc_pkcs15init_skeyargs for proper secret key support

fabled committed:
pkcs15-init: include secret keys in generic object handling code

Type user visible type string is 'secrkey' in harmony with pkcs11-tool.

fabled committed:
pkcs15init: add support for secret key upload and generation

fabled committed:
Support SKDF and secret key upload for MyEID cards

fabled committed:
pkcs15-init: support secret key upload and generation

rickyepoderi committed:
Add dnie_free_apdu_buffers into dnie_sm_free_wrapped_apdu.

rickyepoderi committed:
Use sc_hex_dump instead of cwa_hexdump.

fabled committed:
pkcs15: allow struct sc_pkcs15_prkey to contain secret key

This allows using the existing store_key abstraction to upload
secret keys too.

fabled committed:
add algorithm references for AES keys

frankmorgner committed:
Fix missing error handling of memory allocation (#1020)

* libopensc: handle allocation errors
* handle more faults during memory allocation

fixes several situations that cause segmentation fault

mouse07410 committed:
Fix bug in pkcs11-tool.c that prevented it from displaying CKA_DERIVE key usage for EC public keys

fabled committed:
pkcs15: fix commit 76d59158 auth object regressions

Few conversions of SC_PKCS15_TYPE_AUTH_PIN to SC_PKCS15_TYPE_AUTH
were missed.

fabled committed:
build fix for libressl 2.5.3

X509_up_ref is implemented in libressl 2.5.3

mouse07410 committed:
Fix CAC-introduced bug that broke access to certificates (@dengert found it) (#1032)

Regression from https://github.com/OpenSC/OpenSC/commit/777e2a3751e3f6d53f056c98e9e20e42af674fb1

frankmorgner committed:
Force T=1 for contactless cards

Communication defined by ISO/IEC 14443 is identical to T=1, so make
sure we connect in the right mode to the card so that the constructed
APDUs can be handled by the card.

Jakuje committed:
Restore the functionality of CAC usage flags deriving from X509 certificates

Jakuje committed:
Do not use the hardcoded magic numbers in PIV

fabled committed:
pkcs15init: use random id for secret key

Calculating intrinsic key would probably be not wise, because
it would leak out information about the secret key. Try to
generate globally unique IDs just by using a random one.

frankmorgner committed:
Fixed ssh's pkcs11 usage on macOS

copy (instead of link) our pkcs11 libraries to the default location, which is whitelisted for ssh usage

fixes https://github.com/OpenSC/OpenSC/issues/1008

frankmorgner committed:
sc-hsm: fixed forcing a card driver via opensc.conf

- don't use private data on card matching
- instead, return 1 for every known ATR and only select the applet if the ATR is unknown.
- card initialization always selects the applet.

Advantage: decouples memeory management in matching from initializing the card.
Disadvantage: Applet is selected twice in case of an unknown ATR (once for matching and a second time for initializing the card).

Fixes https://github.com/OpenSC/OpenSC/issues/1042

dengert committed:
piv-tool loops when using OpenSSL-1.1

Add code to support OpenSSL initialization correctly when using OpenSSL-1.1

Tested with OpenSSL-1.1.0c and OpenSSL-1.1.0e.

 Changes to be committed:
	modified:   src/tools/piv-tool.c

dengert committed:
gids-tool loops when using OpenSSL-1.1

Add code to support OpenSSL initialization as needed for OpenSSL-1.1

rickyepoderi committed:
Disable the secure channel at logout (#1048)

Fixes https://github.com/OpenSC/OpenSC/issues/1036

frankmorgner committed:
md: allow user defined data for PIN pad dialog

- use UI framework
- timeout progressbar is running backwards
- cancelling is disabled by default
- removes card specific UI strings, use opensc.conf for that instead
- icon can be loaded by file

frankmorgner committed:
macOS: Added support for notifications

- adds framework for user defined message strings
- automatically determine the system's language

frankmorgner committed:
sc-hsm: add GoID ATRs

disables forcing sc-hsm for standard JCOP ATRs, see
https://github.com/OpenSC/OpenSC/commit/b7de5888a0d9012c99e5f624bab1b98671fc7446#commitcomment-22258928

frankmorgner committed:
sc-hsm: fixed using GoID with unknown ATR

Jakuje committed:
If the underlying PKCS#15 structure does not provide label for a cert… (#1045)

frankmorgner committed:
added compatibility with WiX 3.11

frankmorgner committed:
removed BUILD_ON/BUILD_FOR variable

- guess the system architecture based on %PROCESSOR_ARCHITECTURE%
- guess the target system based on %PLATFORM%

frankmorgner committed:
AppVeyor: Updated OpenSSL version

frankmorgner committed:
AppVeyor: cache zlib build

frankmorgner committed:
AppVeyor: cache openpace build

frankmorgner committed:
fixed bad memory access

frankmorgner committed:
prevent integer overflow

frankmorgner committed:
avoid logically dead code

frankmorgner committed:
fixed ressource leak

frankmorgner committed:
Linux: Added support for notifications

konstantinpersidskiy committed:
pkcs11-tool: add GOSTR3410 keypair generation (#997)

* pkcs11-tool: add GOSTR3410 keypair generation
* Refactor gost3410 generation

fabled committed:
pkcs15-tool: add documentation for secret key options

frankmorgner committed:
Revert "pkcs11-tool: always authenticate when pinpad is in use"

This reverts commit 423375c6f85853d359d502a28e676a09e33a0ac1.

Fixes https://github.com/OpenSC/OpenSC/issues/1063

frankmorgner committed:
fixed Coverity issues

Fixes https://github.com/OpenSC/OpenSC/issues/1057

frankmorgner committed:
0.17.0 version bump

frankmorgner committed:
Windows: Added support for notifications

frankmorgner committed:
pkcs11-tool: respect CKA_ALWAYS_AUTHENTICATE

Quoting from PKCS#11:

The CKA_ALWAYS_AUTHENTICATE attribute can be used to force re-authentication (i.e. force the user to provide a PIN) for each use of a private key. “Use” in this case means a cryptographic operation such as sign or decrypt. This attribute may only be set to CK_TRUE when CKA_PRIVATE is also CK_TRUE.

Re-authentication occurs by calling C_Login with userType set to CKU_CONTEXT_SPECIFIC immediately after a cryptographic operation using the key has been initiated (e.g. after C_SignInit).

Closes https://github.com/OpenSC/OpenSC/pull/1066

frankmorgner committed:
updated required PCSC-Lite version

Fixes https://github.com/OpenSC/OpenSC/issues/1065

frankmorgner committed:
added pull request template

moves github specific files into .github

frankmorgner committed:
use a better issue template

removes CONTRIBUTING.md

Jakuje committed:
clang static analyzer reports

tools/pkcs15-tool.c:
	Dead assignment: Value stored to 'c' is never read
tools/pkcs11-tool.c:
	Dead assignment: Value stored to 'n' is never read
	Dead assignment: Value stored to 'rv' is never read
libopensc/card-cac.c:
	Dead assignemnt: Value stored to 'tl_head_len' is never read
	Dead increment: Value stored to 'outp' is never read
common/libpkcs11.c:
	Memory leak in case of C_UnloadModule() fails
libopensc/pkcs15-pubkey.c:
	Potential memory leaks
pkcs11/mechanism.c:
	Potential memory leak
pkcs11/framework-pkcs15.c:
	Potential memory leaks
	Dereference of null pointer
	Dead assignments
tools/sc-hsm-tool.c:
	Function call argument is an uninitialized value
	Dead assignment: Value stored to 'r' is never read
libopensc/card-openpgp.c:
	Dead assignment: ignoring the errors in case of sc_pkcs15_encode_pubkey() failed
libopensc/pkcs15-cac.c:
	Dead assignments: ignoring return values
libopensc/pkcs15-coolkey.c:
	Dead assignments: ignoring return values
libopensc/card-sc-hsm.c:
	Dereference of undefined pointer value: Properly check the file allocation
pkcs11/slot.c:
	Dead assignment
pkcs15init/pkcs15-cflex.c:
	Dereference of null pointer
	Uninitialized argument values

maciejsszmigiero committed:
Add multiple PINs support to minidriver

Many cards need multiple PINs to work correctly since different on-card
keys are secured by different PINs (this is true for for example OpenPGP
card).

Smart Card Minidriver API has supported such cards since version 6.02
(Vista+).

Use the same method as PKCS#11 driver does to discover user and sign PINs,
for consistency.
However, if there is a default container on card we'll make sure that its
PIN is an user PIN and if there is no default container we'll mark the one
with the user PIN as default.

All other PINs securing containers on card are added as next PINs, up to
MD_MAX_PINS.

Use this opportunity to also fix two cases where a pointer-to-DWORD
variable was passed as pointer-to-size_t parameter to
md_dialog_perform_pin_operation() - they are of different size on Win64.

Signed-off-by: Maciej S. Szmigiero <mail@maciej.szmigiero.name>

maciejsszmigiero committed:
Minidriver card reinitialization cleanup

In minidriver before performing a card operation we currently check whether
the supplied card handles have changed.
If they did the card in reader might have been changed so we reinitialize
it.

However, in few places in reinitialization call path an error returned by
some operation would leave the context in an inconsistent state.

So let's walk through this path to make sure that functions there will exit
cleanly if an error happens.

Also, make sure that all card operations that actually do something have
the necessary check call in the first place and also that they all
consistently check whether VENDOR_SPECIFIC pointer is not NULL before
dereferencing it.

This is a cleanup part of "Keep track of card resets by other contexts in
minidriver" (that is, it does not include the actual reset handling code
introduced by that commit), simplified.

Signed-off-by: Maciej S. Szmigiero <mail@maciej.szmigiero.name>

maciejsszmigiero committed:
sc-hsm: fix unused variable warnings in sc_hsm_init()

When building without OpenPACE there are two unused variables in
sc_hsm_init() that cause compiler to emit warnings about them.

Signed-off-by: Maciej S. Szmigiero <mail@maciej.szmigiero.name>

frankmorgner committed:
opensc-notify: notify for card events

... manually or in daemon mode

fixes compilation in MinGW/Travis CI

frankmorgner committed:
card-piv: Fix Thunderbird and SC_PIN_CMD_GET_INFO

Fixes https://github.com/OpenSC/OpenSC/issues/1071
Closes https://github.com/OpenSC/OpenSC/pull/1072

frankmorgner committed:
fixed library name of smm-local

Commit:
Fix maximum CHR length

Commit:
Fix change PIN failure (#1076)

Jakuje committed:
[cac] Correctly select APDU CASE to unbreak get_challenge

Jakuje committed:
[pkcs11-tool] Do not use modLenBytes before validating

frankmorgner committed:
sc-hsm: New PIN commands for GoID 1.0

Commit:
MyEID - T0 protocol fail for derive operation

apdu.le  must be set minimally to shared key length (pubkey_len / 2),
otherwise sc_get_response() does not read derived shared key from card.

Jakuje committed:
[coolkey] Copy labels from certificate objects to the keys

frankmorgner committed:
quick workarounds for notifications

Jakuje committed:
Simplify differences between CardOS 5 versions and unbreak 5.3 signatures (#1080)

* Simplify CardOS 5.0 support (removing explicit 5.3 marker since the behavior should be the same)

* Restore RSA_PKCS signatures functionality

Closes https://github.com/OpenSC/OpenSC/pull/1079

popovec committed:
MyEID support for RAW RSA signature for 2048 bit keys

MyEID does not support RAW RSA signature for 2048 bit key.
(Source:  MyEID reference manual 2.1.4)
This hack uses decipher operation for calculating
RAW 2048 bit signature.

frankmorgner committed:
PIV: workaround for reading public key

closes https://github.com/OpenSC/OpenSC/pull/1069

frankmorgner committed:
Cope with incomplete ASN.1 data in sc_asn1_read_tag

frankmorgner committed:
Added npa-tool.1 to distribution

fixes https://github.com/OpenSC/OpenSC/issues/1093

frankmorgner committed:
sc-hsm: better error messages for optional files

frankmorgner committed:
opensc-notify: use a dedicated GUID on windows

frankmorgner committed:
opensc.conf: added default notification texts

frankmorgner committed:
opensc-conf: Added notification text for GoID

frankmorgner committed:
sc-hsm: removed UTSRCACC100001

hamano committed:
jpki: set user_consent flag

frankmorgner committed:
0.17.0: updated NEWS file

frankmorgner committed:
win32: Dedicated md/pkcs11 installation folders

Don't install non-permanent files (md/pkcs11) to system folder anymore.

Commit:
Fix memory leak #1050

frankmorgner committed:
adapted date in NEWS

frankmorgner committed:
sc-hsm: fixed building with --disable-sm

Closes https://github.com/OpenSC/OpenSC/pull/1103

frankmorgner committed:
fixed date of 0.17.0 in NEWS

Fixes https://github.com/OpenSC/OpenSC/issues/1105

https://github.com/OpenSC/OpenSC/releases/download/0.17.0/opensc-0.17.0.tar.gz
has been updated to include this fix.

frankmorgner committed:
AppVeyor: cache cngsdk.msi

frankmorgner committed:
build against OpenSSL 1.1.0 on Windows, by default

- build with AppVeyor's OpenSSL 1.1.0
- updates AppVeyor version of OpenPACE

Fixes https://github.com/OpenSC/OpenSC/issues/1108

frankmorgner committed:
AppVeyor: Drop VS2010 build

The pre-compiled OpenSSL binaries on AppVeyor are incompatible with
VS2010

hamano committed:
add public keys

frankmorgner committed:
reader-pcsc: allow no attached reader for notification

Return an error only on OS X, where PnP notification is not supported.

frankmorgner committed:
opensc-notify: fixed heavy load on OS X if no reader

Also, if too many errors are supported, opensc-notify will abort

hamano committed:
documentation for ignore_private_certificate parameter.

frankmorgner committed:
win32: Load notification icons from Windows DLLs

fixes showing wrong icons when loaded as PKCS#11 module

frankmorgner committed:
win32: switch back to OpenSSL 1.0.2 as default

AppVeyor changed the default version back to 1.0.2, see
https://github.com/appveyor/ci/issues/1671

frankmorgner committed:
travis: fixed disabling of notifications for mingw

frankmorgner committed:
fixed compiler warnings

frankmorgner committed:
reader-pcsc: use sc_apdu2bytes for PIN APDU

frankmorgner committed:
win32: fixed OpenSC major/minor version

Jakuje committed:
Do not call list_init() twice (amend 6bfb39454)

frankmorgner committed:
fixed dead assignments

frankmorgner committed:
fixed uninitialized buffer

frankmorgner committed:
fixed potential memory leak

Jakuje committed:
CAC issues (#1120)

* cac: Make the retransmitted APDU valid by restoring the resplen

* cac: Check SWs for all the APDUs and report the errors to underlying layers

* cac: Fallback from CACv1 to CACv2 when CACv1 instruction is not recognized

for the lack of other pointers how to recongnize them

* avoid goto

Jakuje committed:
[pkcs11-tool] Properly define RSA-PSS mechanisms

Jakuje committed:
[pkcs11-tool] Add SHA2 HMAC mechanisms

frankmorgner committed:
sc_read_binary may return less bytes than requested

Note, that there are a number of card drivers that still use
`sc_read_binary` in the wrong way. Unfortunately, I don't have the time
to go through all of them.

Fixes https://github.com/OpenSC/OpenSC/issues/1112

Jakuje committed:
Use shorter PIN name for default PIN to accomodate Card Holder name in future

Jakuje committed:
Get cardholder name from the first certificate

Jakuje committed:
Do not add non-informative PIN to the token label

af-anssi committed:
Support for new MinInt agent card (#1092)

* Support for new MinInt agent card

This card uses the same ATR as the existing card, but the applet installed
does not have the same AID. This card actually works exactly as the
IASECC_SAGEM.

Unify iasecc_init for AMOS/SAGEM and MI cards

Jakuje committed:
The cardos driver supports also CardOS 5

CardOS M4 is confusing since it refers to the old card versions

Jakuje committed:
Parse the ECC parameters from TokenInfo (CHOICE in AlgorithmInfo)

Fixes #1134

Jakuje committed:
Add newly reported ATR for CardOS 5.3

Jakuje committed:
List human-readable version in the cardos-tool

mouse07410 committed:
Fix typo in ECDSA-SHA384 for pkcs11-tool

mouse07410 committed:
Fix bug in pkcs11-tool (did not ASN.1-encode ECDSA-SHA-2 signature)

Commit:
Use |, not ||, when decoding CAC key_reference

This quiets two compiler warnings and is almost surely what was actually
meant.  However, I do not have a CAC card to test.

Commit:
sc_pkcs15_get_bitstring_extension: int, not long long

Use the ASN.1 decoder's SC_ASN1_BIT_FIELD decoder to properly decode
into a machine word.  As _bitstring_extension is used only for the OID
2.5.29.15 by all callers, which is at most 9 bits wide, this is a
reasonable thing to do.

vletoux committed:
Fix wrong padding for GIDS admin authentication

FeitianSmartcardReader committed:
Add ECC support and solve wrong Length status codes with SM card

Fix #1073
Fix #1115

frankmorgner committed:
starcos: added serial number for 3.4

cherezov committed:
sensitive flag added to generate key with CKA_SENSITIVE flag

cherezov committed:
sensitive flag added to generate key with CKA_SENSITIVE flag.
pull request issues fix: replace spaces with tabs and fix description of the argument.

frankmorgner committed:
removed unused test

test is performed by sc_asn1_read_tag

frankmorgner committed:
unified reading of EF.GDO

Jakuje committed:
Do not overwrite digestinfo + hash for RSA-PKCS Signature

Jakuje committed:
Print well formated report from decryption stage and do not fail with OAEP mechanisms

frankmorgner committed:
Added tool for German eGK

Fixes missing opensc-notify.1

frankmorgner committed:
Added PKCS#15 emulator for DIN 66291 profile

vletoux committed:
Gids verified mutual auth

frankmorgner committed:
check for digits in case of BCD encoded PIN

frankmorgner committed:
fixed documentation of GLP encoded PIN

frankmorgner committed:
starcos: 3.4 supports ISO based PIN status queries

Jakuje committed:
Enable RSA-PSS signatures in pkcs11-tool (#1146)

* Add missing SHA224 RSA algorithms

* Fix wrong replacement in pkcs11-tool manual page

* Add MGF and PSS_PARAMS definitions in PKCS#11 header file

* Inspect PSS signature parameters in pkcs11-spy

* Enable RSA-PSS signatures in pkcs11-tool

* Added short names to RSA-PSS methods

* Reintroduce portable NORETURN indication for functions and use it to avoid compilers complaining

frankmorgner committed:
iso7816: don't show error for PIN status on 63C3

dengert committed:
Make OpenSC PKCS#11 Vendor Defined attributes, mechanisms etc unique

This will help when p11-kit is usead and wil allow for additional
CK*_* things to be defined that have a much better chance of being
unique.

OR in "OSC" to any CK*_VENDOR_DEFINED thing.
with #define SC_VENDOR_DEFINED 0x4F534300  /* OSC */

This follows Netscapes convention of doing the same but
using: #define NSSCK_VENDOR_NSS 0x4E534350 /* NSCP */

The current 2 defines  CKA_* are for internal attributes.

 On branch OSC_VENDOR_DEFINED
 Changes to be committed:
	modified:   pkcs11-opensc.h

frankmorgner committed:
fixed building without SM

fixes https://github.com/OpenSC/OpenSC/issues/1155

frankmorgner committed:
CTK: fixed initializing ATR and UID

frankmorgner committed:
export sc_card_find_ec_alg

frankmorgner committed:
CTX: detect card resets

frankmorgner committed:
CTK: Added support for PIN pad

frankmorgner committed:
MacOS: optionally use CTK in package builder

frankmorgner committed:
md: added CNS ATR (#1166)

fixes https://github.com/OpenSC/OpenSC/issues/1153

frankmorgner committed:
Merge remote-tracking branch 'upstream/master' into winui

frankmorgner committed:
md: added default text for role: user+sign

frankmorgner committed:
Merge pull request #1126 from frankmorgner/winui

UI improvements: Customizable Notifications and PIN-Pad Dialog

frankmorgner committed:
added boilerplate for disabling old card driver

currently disabled:
- miocos
- jcop

mouse07410 committed:
Add support in pkcs11-tool and pkcs11-spy for RSA-OAEP (#1169)

* Add RSA-OAEP documentation
* Add definitions and structures needed for RSA-OAEP
* Add ability to display RSA-OAEP parameters to SPY

Closes https://github.com/OpenSC/OpenSC/issues/1170

frankmorgner committed:
added missing include

frankmorgner committed:
msvc: don't warn about deprecated crt functions

See https://msdn.microsoft.com/de-de/library/ms235384.aspx

frankmorgner committed:
fixed unused dependency

frankmorgner committed:
win32: don't link static libs into static libs

frankmorgner committed:
fixed binary comparison

metsma committed:
EstEID ECDSA token support (#1158)

frankmorgner committed:
msvc: disable type truncation warnings

frankmorgner committed:
treat warnings as errors

frankmorgner committed:
removed unused compile options

frankmorgner committed:
fixed dereferencing type-punned pointer

frankmorgner committed:
allow static linking of OpenSCToken

frankmorgner committed:
fixed possible NULL dereference

frankmorgner committed:
fixed possible bad memory access

frankmorgner committed:
cope with system's <strings.h>

fixes https://github.com/OpenSC/OpenSC/issues/1180

frankmorgner committed:
include branch in CI links

frankmorgner committed:
honor return values of read and sscanf

frankmorgner committed:
avoid integer underflow

frankmorgner committed:
fixed dead assignment

frankmorgner committed:
fixed memory leak

frankmorgner committed:
fixed accessing garbage value

frankmorgner committed:
created opensc-asn1

fixes installation of egk-tool.exe on win32

frankmorgner committed:
use pluginkit to register OpenSCToken

frankmorgner committed:
fixed codesigning of opensc tools

frankmorgner committed:
use MACOSX_DEPLOYMENT_TARGET

Jakuje committed:
Enable CAC ALT token card operations

frankmorgner committed:
macOS: Disable notifications by default

fixes https://github.com/OpenSC/OpenSC/issues/1174

Jakuje committed:
Add the ACA path to the PIN structure if we have one

Jakuje committed:
Avoid potential memory leak

frankmorgner committed:
opensc-notify: register a desktop menu entry

- fixes showing notifications in gnome-shell via gio2
- removes gdbus interface for notifications
- fixes https://github.com/OpenSC/OpenSC/issues/1186

frankmorgner committed:
initialize notifications in PKCS#11 and tools

... and don't panic when they are not initialized

metsma committed:
Fix crash when certificate read failed (#1189)

Fixes https://github.com/OpenSC/OpenSC/issues/1176

metsma committed:
Cleanup expired EstEID card ATR-s

Signed-off-by: Raul Metsma <raul@metsma.ee>

metsma committed:
EstEID ECDH token support (#1185)

frankmorgner committed:
beautify printed ASN.1 output

- fixed printing tags on multiple bytes
- align indenting with raw tags
- use OpenSSL's human readable OID database
- only print the canonical names for universal tags

frankmorgner committed:
dump data with sc_debug_hex and sc_log_hex

frankmorgner committed:
print ASN.1 time and ASCII tranlation

frankmorgner committed:
md: fixed memory leak

may be related to https://github.com/OpenSC/OpenSC/issues/1106

frankmorgner committed:
only set MACOSX_DEPLOYMENT_TARGET

which is equivalent to use -mmacosx-version-min

frankmorgner committed:
win32: generate PDB files for releases

https://msdn.microsoft.com/fsk896zz.aspx

fixes https://github.com/OpenSC/OpenSC/issues/1191

metsma committed:
Fix reading EstEID certificates with T=0 (#1193)

Jakuje committed:
Avoid GCC 7 warnings with -Werror (#1196)

* Avoid GCC 7 warnings with -Werror

-Werror=implicit-fallthrough=
	libopensc/card-incrypto34.c
		not sure if this is a bug or intention
	libopensc/card-rutoken.c
		most probably intention
	libopensc/card-westcos.c
		remove bogus if so the compile is not confused
		I will fill a separate bug to gcc probably
	pkcs15init/pkcs15-iasecc.c
		Simplify the log and avoid compiler confusion
	sm/sm-common.c
		explicit fallthrough
	tools/pkcs11-tool.c
		use explicit fallthrough comment
	tools/pkcs15-init.c
		The fallthrough is obvious here

-Werror=format-truncation=
	libopensc/pkcs15-itacns.c
		use explicit string lengths
	pkcs11/framework-pkcs15.c
		calculate the truncation
	tests/pintest.c
		avoid sprintf
	tools/pkcs15-crypt.c
		avoid sprintf
	tools/pkcs15-init.c
		calculate the truncation

frankmorgner committed:
added missing opensc-asn1.1

fixes https://github.com/OpenSC/OpenSC/issues/1206

frankmorgner committed:
md: protect MD entry points with CriticalSection

CardDeleteContext may be called at any time, interrupting any ongoing
operation with the same PCARD_DATA. This leads to a race condition when
CardDeleteContext deletes, for example, the sc_context_t which the
interrupted call still wants to access. We have seen and fixed this
problem in https://github.com/OpenSC/OpenSC/issues/973 specifically for
the PIN entry process, however, it also applies to all other calls to
the md.

The new implementation removes the need for global data in the md.

frankmorgner committed:
opensc-explorer/opensc-tool: Allow extended APDUs

frankmorgner committed:
ISO 7816-4 doesn't force a maximum for GET CHALLENGE

frankmorgner committed:
fixed compiler warning

frankmorgner committed:
CardOS: Try forcing max_send_size for PSO:DEC

Fixes https://github.com/OpenSC/OpenSC/issues/1208
Fixes https://github.com/OpenSC/OpenSC/issues/1118
Fixes https://github.com/OpenSC/OpenSC/issues/1005
Fixes https://github.com/OpenSC/OpenSC/issues/802

frankmorgner committed:
SC-HSM: Don't block generic contactless ATR

frankmorgner committed:
sc-hsm: check integrity of SM channel

cope with multiple processes that may reset the SM channel when
accessing the card at the same time

frankmorgner committed:
macOS High Sierra: fixed dmg creation

fixes https://github.com/OpenSC/OpenSC/issues/1198

frankmorgner committed:
macOS: fixed detection of OpenPACE package

frankmorgner committed:
OpenPGP: Added basic support for OpenPGP card V3

partially implements https://github.com/OpenSC/OpenSC/issues/1215

Refactored OpenPGP code so that future versions of the card will be
accessed using the logic for OpenPGP V2. We hope that backward
compatibility of the standard will keep the new versions functional.

frankmorgner committed:
OpenPGP: fixed selecting Applet

fixes https://github.com/OpenSC/OpenSC/issues/1230

frankmorgner committed:
fixed warnings about uninitialized data

frankmorgner committed:
ISO7816: allow nested CP DOs

frankmorgner committed:
openpgp: factory reset is possible if LCS is supported

frankmorgner committed:
IAS/ECC: fixed applet selection

fixes https://github.com/OpenSC/OpenSC/issues/1240

frankmorgner committed:
PC/SC: don't reset the card on disconnection

Windows/macOS (minidriver/tokend) handle the authentication status and
perform an explicit logout on shutdown. PKCS#11 standard requires a
session for logging into the card; when closing the session we perform
an explicit logout. Hence, the authentication status should be reset
even if not performing a reset on disconnect.

Jakuje committed:
SimpleTLV: Skip correctly two bytes after reading 2b size (#1231)

frankmorgner committed:
OpenPGP: Added support for PIN logout and status

frankmorgner committed:
OpenPGP: implemented GET CHALLENGE

frankmorgner committed:
documented usage of SM_MODE_ACL

fixes https://github.com/OpenSC/OpenSC/issues/971

Mardalemer committed:
pkcs15init: Fix rutokenS FCP parsing

RutokenS returns data with little endian byte order, due to this
fact token wouldn't work with standard function. So function for
parsing fcp from little endian data was inplemented.

dfortinskiy1 commented:
merge, add commit message and lgtm

lolyonok commented on src/libopensc/card-rutoken.c:
```diff
@@ -452,6 +442,99 @@ static int rutoken_select_file(sc_card_t *card,
+	/* determine file descriptor */
+	tag = sc_asn1_find_tag(card->ctx, buf, buflen, 0x82, &taglen);
+	if (tag != NULL && taglen == 2) {
+		/* Here means only leading byte. */
```
Only leading byte is significant

lolyonok replied:
надо определиться комменты с большой или маленькой буквы

lolyonok commented on src/libopensc/card-rutoken.c:
```diff
@@ -452,6 +442,99 @@ static int rutoken_select_file(sc_card_t *card,
+		/* Here means only leading byte. */
+		const char* type;
+
+		file->shareable = tag[0] & 0x40 ? 1 : 0;
```
нужна ли проверка? или у нас всегда 0?

lolyonok commented on src/libopensc/card-rutoken.c:
```diff
@@ -452,6 +442,99 @@ static int rutoken_select_file(sc_card_t *card,
+	/* determine RSF information */
+	tag = sc_asn1_find_tag(card->ctx, buf, buflen, 0x85, &taglen);
+	if (tag != NULL && taglen > 0) {
+		sc_file_set_prop_attr(file, tag, taglen);
```
скобки вроде тут не ставятся на однострочные if

lolyonok commented on src/libopensc/card-rutoken.c:
```diff
@@ -452,6 +442,99 @@ static int rutoken_select_file(sc_card_t *card,
+	}
+
+	/* determine file lifecycle */
+	tag = sc_asn1_find_tag(card->ctx, buf, buflen, 0x8A, &taglen);
```
у нас используется этот тэг?
бывают ли у нас неактивированные файлы

frankmorgner committed:
DNIe: card also supports 1920 bits (#1247)

fixes https://github.com/OpenSC/OpenSC/issues/1246

edgarholleis committed:
Fix memleak (#1255)

lolyonok commented on src/libopensc/card-rutoken.c:
```diff
@@ -452,6 +442,95 @@ static int rutoken_select_file(sc_card_t *card,
+
+	/* Determine file id. */
+	tag = sc_asn1_find_tag(card->ctx, buf, buflen, 0x83, &taglen);
+	if (tag != NULL && taglen == 2) {
```
открывающие скобки с новой строки

lolyonok commented on src/libopensc/card-rutoken.c:
```diff
@@ -452,6 +442,95 @@ static int rutoken_select_file(sc_card_t *card,
+
+	/* Determine file lifecycle. */
+	tag = sc_asn1_find_tag(card->ctx, buf, buflen, 0x8A, &taglen);
+	if (tag != NULL && taglen == 1)
```
стоит поставить скобки

lolyonok commented:
схлопнуть в один коммит и lgtm

Mardalemer committed:
pkcs15init: Fix rutokenS FCP parsing

RutokenS returns data with little endian byte order, due to this
fact token wouldn't work with standard function. So function for
parsing fcp from little endian data was inplemented.
